rust/src/push/evaluator.rs

```diff
@@ -105,6 +105,17 @@ impl PushRuleEvaluator {
     /// Create a new `PushRuleEvaluator`. See struct docstring for details.
     #[allow(clippy::too_many_arguments)]
     #[new]
+    #[pyo3(signature = (
```

Comment:
any chance for an explanation on what this is doing or why it's needed?

Reply:
New Pyo3 versions require that an explicit signature is given if any but the last parameter type is an `Option`. See https://pyo3.rs/main/function/signature#trailing-optional-arguments.

Reply:
If not used, clippy fails with a [deprectation message](https://github.com/matrix-org/synapse/actions/runs/5943525031/job/16118845691#step:5:26) that reads:

```
 error: use of deprecated constant `pyo3::impl_::deprecations::REQUIRED_ARGUMENT_AFTER_OPTION`: required arguments after an `Option<_>` argument are ambiguous and being phased out
       = help: add a `#[pyo3(signature)]` annotation on this function to unambiguously specify the default values for all optional parameters
   --> rust/src/push/evaluator.rs:113:9
    |
113 |         notification_power_levels: BTreeMap<String, i64>,
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `-D deprecated` implied by `-D warnings`
```